File: legacy-swift/Sources/LLMCostMonitor/Infrastructure/Pricing/PricingTable.swift
import Foundation

/// Loader for the vendored LiteLLM `model_prices_and_context_window.json` snapshot.
/// Refreshed by `Scripts/refresh-pricing.sh`. We deliberately do NOT decode the whole file
/// into typed structs — the LiteLLM schema is gnarly (model entries have wildly varying
/// optional keys per provider), so we keep raw JSON and pull only the fields we need.
public struct PricingTable: Sendable {
    /// Per-million-token prices for one model. All values are USD per 1M tokens.
    public struct ModelPrice: Sendable, Equatable {
        public let inputPerM: Decimal
        public let outputPerM: Decimal
        public let cacheReadPerM: Decimal?
        public let cacheCreationPerM: Decimal?
        public let cacheCreation5mPerM: Decimal?
        public let cacheCreation1hPerM: Decimal?
        public let reasoningPerM: Decimal?
        public let contextWindow: Int?
        public let provider: String?
    }

    /// SHA prefix of the source `pricing.json` — recorded on each `UsageEvent`.
    public let snapshotVersion: String

    private let entries: [String: ModelPrice]

    /// Load from the bundled `Resources/pricing.json`.
    public static func loadBundled() throws -> PricingTable {
        guard let url = Bundle.module.url(forResource: "pricing", withExtension: "json") else {
            throw PricingError.bundleResourceMissing
        }
        return try load(from: url)
    }

    public static func load(from url: URL) throws -> PricingTable {
        let data = try Data(contentsOf: url)
        return try load(from: data)
    }

    public static func load(from data: Data) throws -> PricingTable {
        guard let root = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw PricingError.invalidRoot
        }

        var entries: [String: ModelPrice] = [:]
        entries.reserveCapacity(root.count)

        for (key, value) in root {
            guard let dict = value as? [String: Any] else { continue }
            // LiteLLM's first entry is "sample_spec" — skip non-priced entries.
            let inputCost = dict.decimalValue(forKey: "input_cost_per_token")
            let outputCost = dict.decimalValue(forKey: "output_cost_per_token")
            guard inputCost != nil || outputCost != nil else { continue }

            let cacheRead = dict.decimalValue(forKey: "cache_read_input_token_cost")
            let cacheCreation = dict.decimalValue(forKey: "cache_creation_input_token_cost")
            let cache5m = dict.decimalValue(forKey: "cache_creation_input_token_cost_above_5m")
                ?? dict.decimalValue(forKey: "cache_creation_5m_input_token_cost")
            let cache1h = dict.decimalValue(forKey: "cache_creation_input_token_cost_above_1hr")
                ?? dict.decimalValue(forKey: "cache_creation_1h_input_token_cost")
            let reasoning = dict.decimalValue(forKey: "output_reasoning_token_cost")

            let cw = (dict["max_input_tokens"] as? Int) ?? (dict["max_tokens"] as? Int)

            entries[key.lowercased()] = ModelPrice(
                inputPerM:           (inputCost ?? 0) * 1_000_000,
                outputPerM:          (outputCost ?? 0) * 1_000_000,
                cacheReadPerM:       cacheRead.map { $0 * 1_000_000 },
                cacheCreationPerM:   cacheCreation.map { $0 * 1_000_000 },
                cacheCreation5mPerM: cache5m.map { $0 * 1_000_000 },
                cacheCreation1hPerM: cache1h.map { $0 * 1_000_000 },
                reasoningPerM:       reasoning.map { $0 * 1_000_000 },
                contextWindow:       cw,
                provider:            dict["litellm_provider"] as? String
            )
        }

        // Snapshot fingerprint = SHA-256 prefix of the bytes. Cheap, stable.
        let version = ShortHash.sha256Prefix(data, length: 12)

        return PricingTable(snapshotVersion: version, entries: entries)
    }

    public init(snapshotVersion: String, entries: [String: ModelPrice]) {
        self.snapshotVersion = snapshotVersion
        self.entries = entries
    }

    /// Look up pricing for `model`. Tries:
    /// 1. exact (case-insensitive)
    /// 2. provider-prefixed match (`anthropic/<model>`)
    /// 3. prefix scan (`<entry>` is a prefix of `model` or vice-versa)
    /// 4. keyword fallback (`opus`/`sonnet`/`haiku`/`gpt`/`gemini`/...)
    /// 5. nil
    public func price(for model: String) -> ModelPrice? {
        let lowered = model.lowercased()
        if let direct = entries[lowered] { return direct }

        // Try common provider-prefixed forms.
        for prefix in ["anthropic/", "openai/", "google/", "vertex_ai/", "bedrock/"] {
            if let hit = entries[prefix + lowered] { return hit }
        }

        // Substring sweep — first stable match wins.
        for (key, price) in entries {
            if lowered.hasPrefix(key) || key.hasPrefix(lowered) { return price }
        }

        // Keyword fallback. Returns the *first* model entry whose key contains the keyword,
        // and where pricing is non-zero — keeps us from picking a deprecated zero-priced row.
        let keywords = ["opus", "sonnet", "haiku", "gpt", "gemini", "o1", "o3", "o4",
                        "kimi", "qwen", "deepseek", "glm", "grok", "mistral", "mixtral", "llama"]
        for kw in keywords where lowered.contains(kw) {
            if let hit = entries.first(where: { $0.key.contains(kw) && $0.value.outputPerM > 0 })?.value {
                return hit
            }
        }
        return nil
    }

    /// Compute USD cost for an event using this snapshot.
    /// Falls back to default Sonnet-equivalent pricing if model unknown (so we never silently emit $0).
    public func cost(for event: UsageEvent) -> Decimal {
        let p = price(for: event.model) ?? ModelPrice.fallback
        let perToken: (Int, Decimal) -> Decimal = { tokens, perM in
            guard tokens > 0, perM > 0 else { return 0 }
            return Decimal(tokens) * perM / 1_000_000
        }

        let cache5m = p.cacheCreation5mPerM ?? p.cacheCreationPerM ?? 0
        let cache1h = p.cacheCreation1hPerM ?? p.cacheCreationPerM ?? 0
        let cacheRead = p.cacheReadPerM ?? 0

        let input    = perToken(event.inputTokens, p.inputPerM)
        let output   = perToken(event.outputTokens, p.outputPerM)
        let crRead   = perToken(event.cacheReadTokens, cacheRead)
        let cr5      = perToken(event.cacheCreation5mTokens, cache5m)
        let cr1      = perToken(event.cacheCreation1hTokens, cache1h)
        let reason   = perToken(event.reasoningTokens ?? 0, p.reasoningPerM ?? p.outputPerM)

        return input + output + crRead + cr5 + cr1 + reason
    }

    public func contextWindow(for model: String) -> Int? {
        price(for: model)?.contextWindow
    }

    public var modelCount: Int { entries.count }
}

public enum PricingError: Error, Equatable {
    case bundleResourceMissing
    case invalidRoot
}

private extension PricingTable.ModelPrice {
    /// Sonnet-tier defaults, used only when a model is completely unknown.
    /// Better to over-estimate than to silently emit $0 and make the user think coding is free.
    static let fallback = PricingTable.ModelPrice(
        inputPerM: 3, outputPerM: 15,
        cacheReadPerM: 0.30, cacheCreationPerM: 3.75,
        cacheCreation5mPerM: 3.75, cacheCreation1hPerM: 6.00,
        reasoningPerM: nil,
        contextWindow: 200_000,
        provider: "fallback"
    )
}

private extension Dictionary where Key == String, Value == Any {
    /// LiteLLM stores numbers as either `Int`, `Double`, or string. Decode to `Decimal` losslessly.
    func decimalValue(forKey key: String) -> Decimal? {
        guard let raw = self[key] else { return nil }
        if let d = raw as? Double { return Decimal(d) }
        if let i = raw as? Int { return Decimal(i) }
        if let s = raw as? String, let d = Decimal(string: s) { return d }
        if let n = raw as? NSNumber { return n.decimalValue }
        return nil
    }
}

File: src/main/pricing/pricing-table.ts
import { createHash } from 'node:crypto'
import { readFileSync } from 'node:fs'

import type { UsageEvent } from '@shared/usage-event'

export interface ModelPrice {
  // All values are USD per 1M tokens, stored as float (range fits cleanly).
  readonly inputPerM: number
  readonly outputPerM: number
  readonly cacheReadPerM: number | null
  readonly cacheCreationPerM: number | null
  readonly cacheCreation5mPerM: number | null
  readonly cacheCreation1hPerM: number | null
  readonly reasoningPerM: number | null
  readonly contextWindow: number | null
  readonly provider: string | null
}

// Sonnet-tier defaults, used only when a model is completely unknown.
// Better to over-estimate than silently emit $0 and let the user think coding is free.
const FALLBACK: ModelPrice = {
  inputPerM: 3,
  outputPerM: 15,
  cacheReadPerM: 0.3,
  cacheCreationPerM: 3.75,
  cacheCreation5mPerM: 3.75,
  cacheCreation1hPerM: 6.0,
  reasoningPerM: null,
  contextWindow: 200_000,
  provider: 'fallback',
}

// Keys we keep from each LiteLLM entry. Same set as the Swift port.
const PROVIDER_PREFIXES = ['anthropic/', 'openai/', 'google/', 'vertex_ai/', 'bedrock/']
const LOCAL_PROVIDERS = new Set(['local', 'ollama', 'lmstudio', 'lm_studio', 'llama_cpp', 'llamacpp'])
const SUPPLEMENTAL_PREFIXES = [
  'anthropic.',
  'global.anthropic.',
  'us.anthropic.',
  'eu.anthropic.',
  'au.anthropic.',
  ...PROVIDER_PREFIXES,
]
const KEYWORDS = [
  'opus',
  'sonnet',
  'haiku',
  'gpt',
  'gemini',
  'o1',
  'o3',
  'o4',
  'kimi',
  'qwen',
  'deepseek',
  'glm',
  'grok',
  'mistral',
  'mixtral',
  'llama',
]

function readNumber(raw: unknown): number | null {
  if (typeof raw === 'number' && Number.isFinite(raw)) return raw
  if (typeof raw === 'string') {
    const n = Number(raw)
    return Number.isFinite(n) ? n : null
  }
  return null
}

function buildEntries(root: Record<string, unknown>): Map<string, ModelPrice> {
  const out = new Map<string, ModelPrice>()
  for (const [key, value] of Object.entries(root)) {
    if (value === null || typeof value !== 'object') continue
    const dict = value as Record<string, unknown>

    const inputCost = readNumber(dict['input_cost_per_token'])
    const outputCost = readNumber(dict['output_cost_per_token'])
    // LiteLLM's first entry is `sample_spec` — has no real prices; skip it (and
    // any other entry that lacks both input and output cost).
    if (inputCost === null && outputCost === null) continue

    const cacheRead = readNumber(dict['cache_read_input_token_cost'])
    const cacheCreation = readNumber(dict['cache_creation_input_token_cost'])
    const cache5m =
      readNumber(dict['cache_creation_input_token_cost_above_5m']) ??
      readNumber(dict['cache_creation_5m_input_token_cost'])
    const cache1h =
      readNumber(dict['cache_creation_input_token_cost_above_1hr']) ??
      readNumber(dict['cache_creation_1h_input_token_cost'])
    const reasoning = readNumber(dict['output_reasoning_token_cost'])

    const cw = readNumber(dict['max_input_tokens']) ?? readNumber(dict['max_tokens'])
    const provider = typeof dict['litellm_provider'] === 'string' ? (dict['litellm_provider'] as string) : null

    out.set(key.toLowerCase(), {
      inputPerM: (inputCost ?? 0) * 1_000_000,
      outputPerM: (outputCost ?? 0) * 1_000_000,
      cacheReadPerM: cacheRead === null ? null : cacheRead * 1_000_000,
      cacheCreationPerM: cacheCreation === null ? null : cacheCreation * 1_000_000,
      cacheCreation5mPerM: cache5m === null ? null : cache5m * 1_000_000,
      cacheCreation1hPerM: cache1h === null ? null : cache1h * 1_000_000,
      reasoningPerM: reasoning === null ? null : reasoning * 1_000_000,
      contextWindow: cw === null ? null : Math.trunc(cw),
      provider,
    })
  }
  return out
}

function tokensTimesPerMillion(tokens: number, perMillion: number): bigint {
  if (tokens <= 0 || perMillion <= 0) return 0n
  // tokens × (USD per 1M) = micro-USD, since (1M tokens × USD/M) ÷ 1M = USD,
  // i.e. tokens × USD/M ≡ tokens × USD/M (already in micro-USD because USD/M is
  // pre-scaled by 1e6). Round to integer micro-USD at the boundary.
  return BigInt(Math.round(tokens * perMillion))
}

function mergeMissingPrice(base: ModelPrice, supplement: ModelPrice): ModelPrice {
  return {
    inputPerM: base.inputPerM,
    outputPerM: base.outputPerM,
    cacheReadPerM: base.cacheReadPerM ?? supplement.cacheReadPerM,
    cacheCreationPerM: base.cacheCreationPerM ?? supplement.cacheCreationPerM,
    cacheCreation5mPerM: base.cacheCreation5mPerM ?? supplement.cacheCreation5mPerM,
    cacheCreation1hPerM: base.cacheCreation1hPerM ?? supplement.cacheCreation1hPerM,
    reasoningPerM: base.reasoningPerM ?? supplement.reasoningPerM,
    contextWindow: base.contextWindow ?? supplement.contextWindow,
    provider: base.provider ?? supplement.provider,
  }
}

export class PricingTable {
  readonly snapshotVersion: string
  private readonly entries: Map<string, ModelPrice>

  constructor(snapshotVersion: string, entries: Map<string, ModelPrice>) {
    this.snapshotVersion = snapshotVersion
    this.entries = entries
  }

  get modelCount(): number {
    return this.entries.size
  }

  static fromBuffer(buf: Buffer): PricingTable {
    const text = buf.toString('utf8')
    const root: unknown = JSON.parse(text)
    if (root === null || typeof root !== 'object' || Array.isArray(root)) {
      throw new Error('PricingTable: invalid root (expected JSON object)')
    }
    const entries = buildEntries(root as Record<string, unknown>)
    const version = createHash('sha256').update(buf).digest('hex').slice(0, 12)
    return new PricingTable(version, entries)
  }

  static fromFile(path: string): PricingTable {
    return PricingTable.fromBuffer(readFileSync(path))
  }

  // Lookup chain (from docs/architecture.md D5):
  //  1. exact (case-insensitive)
  //  2. provider-prefixed (anthropic/<model>, openai/<model>, …)
  //  3. prefix sweep (entry-prefix-of-model OR model-prefix-of-entry)
  //  4. keyword fallback (opus|sonnet|haiku|gpt|gemini|o1|o3|o4|kimi|qwen|deepseek|glm|grok|mistral|mixtral|llama)
  //  5. null (caller falls back to FALLBACK in `cost()`)
  price(model: string): ModelPrice | null {
    const lowered = model.toLowerCase()

    const direct = this.entries.get(lowered)
    if (direct !== undefined) return this.withSupplementalAliases(lowered, direct)

    for (const prefix of PROVIDER_PREFIXES) {
      const hit = this.entries.get(prefix + lowered)
      if (hit !== undefined) return hit
    }

    for (const [key, p] of this.entries) {
      if (lowered.startsWith(key) || key.startsWith(lowered)) return p
    }

    for (const kw of KEYWORDS) {
      if (!lowered.includes(kw)) continue
      for (const [key, p] of this.entries) {
        if (key.includes(kw) && p.outputPerM > 0) return p
      }
    }

    return null
  }

  private withSupplementalAliases(model: string, base: ModelPrice): ModelPrice {
    let out = base
    for (const prefix of SUPPLEMENTAL_PREFIXES) {
      const supplement = this.entries.get(prefix + model)
      if (supplement !== undefined) out = mergeMissingPrice(out, supplement)
    }
    return out
  }

  contextWindow(model: string): number | null {
    return this.price(model)?.contextWindow ?? null
  }

  // Compute cost (in integer micro-USD) for an event using this snapshot.
  // Falls back to Sonnet-equivalent pricing if the model is unknown — never
  // silently emits $0. Reasoning tokens reuse outputPerM when no dedicated rate.
  cost(event: UsageEvent): bigint {
    if (LOCAL_PROVIDERS.has(event.provider.toLowerCase())) return 0n

    const p = this.price(event.model) ?? FALLBACK

    const cache5m = p.cacheCreation5mPerM ?? p.cacheCreationPerM ?? 0
    const cache1h = p.cacheCreation1hPerM ?? p.cacheCreationPerM ?? 0
    const cacheRead = p.cacheReadPerM ?? 0
    const reasoningRate = p.reasoningPerM ?? p.outputPerM

    let total = 0n
    total += tokensTimesPerMillion(event.inputTokens, p.inputPerM)
    total += tokensTimesPerMillion(event.outputTokens, p.outputPerM)
    total += tokensTimesPerMillion(event.cacheReadTokens, cacheRead)
    total += tokensTimesPerMillion(event.cacheCreation5mTokens, cache5m)
    total += tokensTimesPerMillion(event.cacheCreation1hTokens, cache1h)
    total += tokensTimesPerMillion(event.reasoningTokens ?? 0, reasoningRate)
    return total
  }
}
